src/drivers/imx/sai.c

```diff
@@ -201,21 +229,20 @@ static inline int sai_set_config(struct dai *dai, struct ipc_config_dai *common_
 	uint32_t val_cr2 = 0, val_cr4 = 0, val_cr5 = 0;
 	uint32_t mask_cr2 = 0, mask_cr4 = 0, mask_cr5 = 0;
 	struct sai_pdata *sai = dai_get_drvdata(dai);
-	/* TODO: this value will be provided by config */
-#ifndef CONFIG_IMX8ULP
-	uint32_t sywd = 32;
-	uint32_t twm = ~(BIT(0) | BIT(1));
-	uint32_t clk_div = SAI_CLOCK_DIV;
-#else
-	uint32_t sywd = 16;
-	uint32_t twm = ~BIT(0);
-	uint32_t clk_div = config->sai.fsync_rate == 8000 ? SAI_CLOCK_DIV : SAI_CLOCK_DIV_16K;
-#endif
 
 	sai->config = *config;
 	sai->params = config->sai;
 
-	val_cr4 |= REG_SAI_CR4_MF;
+	/* bit width of a single slot; FIFO word always 32b */
+	uint32_t sywd = sai->params.tdm_slot_width;
+
+	/*
+	 * Divides the audio main clock to generate the bit clock when
+	 * configured for an internal bit clock.
+	 * The division value is (DIV + 1) * 2.
+	 */
+
+	uint32_t clk_div = (config->sai.mclk_rate / config->sai.bclk_rate / 2) - 1;
```

Comment:
What if `MCLK` and `BCLK` have the same frequency? You'll probably have to bypass the `BCLK` divider using the `BYP` bit from xCR2.